django/db/migrations/graph.py

```diff
@@ -108,27 +131,101 @@ def __init__(self):
         self.nodes = {}
         self.cached = False
 
-    def add_node(self, key, implementation):
-        node = Node(key)
-        self.node_map[key] = node
-        self.nodes[key] = implementation
+    def add_node(self, key, migration):
+        # If the key already exists, then it must be a dummy node.
+        dummy_node = self.node_map.get(key)
+        if dummy_node:
+            # Promote DummyNode to Node.
+            dummy_node.promote()
+        else:
+            node = Node(key)
+            self.node_map[key] = node
+        self.nodes[key] = migration
         self.clear_cache()
 
-    def add_dependency(self, migration, child, parent):
+    def add_dummy_node(self, key, origin, error_message):
+        node = DummyNode(key, origin, error_message)
+        self.node_map[key] = node
+        self.nodes[key] = None
+
+    def add_dependency(self, migration, child, parent, skip_validation=False):
+        """
+        This may create dummy nodes if they don't yet exist.
+        If `skip_validation` is set, validate_consistency should be called afterwards.
+        """
         if child not in self.nodes:
-            raise NodeNotFoundError(
-                "Migration %s dependencies reference nonexistent child node %r" % (migration, child),
-                child
+            error_message = (
+                "Migration %s dependencies reference nonexistent"
+                " child node %r" % (migration, child)
             )
+            self.add_dummy_node(child, migration, error_message)
         if parent not in self.nodes:
-            raise NodeNotFoundError(
-                "Migration %s dependencies reference nonexistent parent node %r" % (migration, parent),
-                parent
+            error_message = (
+                "Migration %s dependencies reference nonexistent"
+                " parent node %r" % (migration, parent)
             )
+            self.add_dummy_node(parent, migration, error_message)
         self.node_map[child].add_parent(self.node_map[parent])
         self.node_map[parent].add_child(self.node_map[child])
+        if not skip_validation:
+            self.validate_consistency()
         self.clear_cache()
 
+    def replace_node(self, key, replacement):
+        """
+        If `key` exists, it is removed and any dependencies that were
+        referencing it are changed to reference the `replacement` node
+        instead.
+        """
+        self.nodes.pop(key, None)
+        node = self.node_map.pop(key, None)
+        if node:
+            replacement_node = self.node_map[replacement]
+            for child in node.children:
+                child.parents.remove(node)
+                if child.key != replacement:
```

Comment:
Yes. To prevent self-referencing.

The replacement node is not aware of the other nodes it is replacing. When there are several nodes for a replacing migration to replace, `build_graph` loops over them, calling `replace_node` several times. After the first node is replaced, the replacement node may become a parent or child of one of the other nodes that it will be replacing. Further `replace_node` calls could then create self-referencing dependencies at the replacement node.

After some thought though, I'm thinking it would be cleaner to not allow the replacement migration to become a parent/child of any of its replacement targets in the first place by turning this function into `replace_nodes` (plural) and giving it all of the replacement targets at once. In this case, before creating new dependencies between the replacement node and some other node, we'd be able to ensure that this other node is not within the set of replacement targets. Do you think this would be cleaner?